### PR TITLE
Update glfw-rs to be on par with GLFW 3.2

### PR DESCRIPTION
**Pull request by novacrazy**
Hello! I had a few days mostly free so I thought I'd work on this project since I'll be using it myself soon enough.

Additions:
* Full GLFW 3.1 and 3.2 support (I think)
* Optional integration with `vk_sys` and `image` for Vulkan and custom Cursor/Icon features, respectively.
* Properly exposed more window hints
    + Context creation API
    + No error mode for contexts
    + Double buffer toggle
    + Context release modes
* Added examples for custom cursors and window icons, and a very, very brief Vulkan example
* Variations of the `with_*` callback functions that can accept an `FnMut` closure. I actually needed this, and if I needed it I'm sure others will, too.

Breaking changes:
* Moved keyboard modifiers to their own module to clarify docs and usage
* For all window hints that could take `GLFW_DONT_CARE`, make them `Option<T>` where `None` sends `DONT_CARE` to `glfwWindowHint`
* Changed `window.set_swap_interval` to take an Enum with the values `None`, `Adaptive` and `Sync(u32)`, so that it can support adaptive sync on platforms that support it. Previously this couldn't be done directly since `set_swap_interval` accepted an unsigned integer, and adaptive sync required `-1` be passed.
* Added variable onto `Window` struct for current `Cursor` instance, so it doesn't `Drop` immediately after passing it. It needed to exist as long as the cursor was in use.

Both Vulkan and `image` library features are behind feature flags and optional dependencies so as to not bloat the library if possible.

If I'm missing anything, please tell me and I'll work on it right away.

### CODE REVIEW

**Comment by novacrazy** [comment]
The tests can't seem to find `image` or `vulkan` features, but the examples run.

**Comment by novacrazy** [comment]
Yay, doctests passed (or ignored) and using only stable features.

File: src/lib.rs [hunk 1]
```diff
@@ -1669,6 +1682,14 @@ impl Window {
         }
     }
 
+    /// Wrapper for `glfwFocusWindow`
+    ///
+    /// It is NOT recommended to use this function, as it steals focus from other applications
+    /// and can be extremely disruptive to the user.
+    pub fn focus(&mut self) {
```
Comment (novacrazy) [inline src/lib.rs]: And `glfwFocusWindow` wrapper. Forgot to mention that.

**Comment by novacrazy** [comment]
So I just noticed the issue on relicensing the project, and my updating the copyright dates of the files I worked on might have been a bit premature. I can revert those if need be.

**Comment by novacrazy** [comment]
Any thoughts on this yet?

**Comment by bvssvni** [comment]
Looks good to me. How does this affect relicensing?

**Comment by novacrazy** [comment]
I usually license all my stuff as MIT, so I'm fine with anything.

What do you think of those breaking changes? It might make more sense to use a custom enum type instead of `Option` for the `DONT_CARE` values, but it doesn't really matter.

**Comment by bvssvni** [comment]
I'm OK with the breaking changes.

**Comment by novacrazy** [comment]
Just added a couple more things really quick that I thought of. I think that's all. 

So if you're good with everything I'll leave it to you.

**Comment by bvssvni** [comment]
Nice!

**Comment by novacrazy** [comment]
Don't forget to update issues #350, #351 and #288. :)

**Pull request merged**